### PR TITLE
chore(qa): refactor text qa to add support for additional models and fix issues

### PR DESCRIPTION
**Pull request by krokoko**
Fixes #

- refactor text qa to add support for new models, including Claude v3
- upgrade packages to get access to new upgrades in langchain
- fix issue where model args were not passed to the selected model
- fix graphql schema in rag preventing both qa and rag to be deployed together (conflicting input type)
- improve logging for image qa

Tested with all models supported, rag and long context, and tested image qa

In a next PR, will also refactor the image_qa to support the new structure and add missing models from Bedrock (mistral,...)

Eventually this code will move to a separate package to be used across the constructs

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the project license.

### CODE REVIEW

**Review by laithalsaadoon — APPROVED** [review]
LGTM

File: lambda/aws-qa-appsync-opensearch/question_answering/src/adapters/bedrock/claude.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+        if self.callback_handler:
+            kwargs["callbacks"] = self.callback_handler
+            kwargs["streaming"] = model_kwargs.get("streaming", False)
+
+        return BedrockChat(
+            **kwargs
+        )
+
+    def get_prompt(self):
```
Comment (dineshSajwan) [inline lambda/aws-qa-appsync-opensearch/question_answering/src/adapters/bedrock/claude.py:115]: Just a heads up when we test it with multimodal, we may need a different prompt for multimodal to include parameter like <Image>.

File: lambda/aws-qa-appsync-opensearch/question_answering/src/adapters/bedrock/claude.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        if "top_p" in model_kwargs:
+            params["top_p"] = model_kwargs["top_p"]
+        if "max_tokens" in model_kwargs:
+            params["max_tokens"] = model_kwargs["max_tokens"]
+        if "stop_sequences" in model_kwargs:
+            params["stop_sequences"] = model_kwargs["stop_sequences"]
+        if "top_k" in model_kwargs:
+            params["top_k"] = model_kwargs["top_k"]
+
+        params["anthropic_version"] = "bedrock-2023-05-31"
```
Comment (dineshSajwan) [inline lambda/aws-qa-appsync-opensearch/question_answering/src/adapters/bedrock/claude.py:98]: can we put the version in a constant type file.

File: lambda/aws-qa-appsync-opensearch/question_answering/src/adapters/bedrock/titan.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        Question: {question}
+
+        Assistant:"""
+
+        return PromptTemplate(
+            template=template, input_variables=["context", "question"]
+        )
+
+# Register the adapter
+registry.register(r"^Bedrock.amazon.titan*", BedrockTitanAdapter)
```
Comment (dineshSajwan) [inline lambda/aws-qa-appsync-opensearch/question_answering/src/adapters/bedrock/titan.py:76]: is this applicable with titan-image- generator as well? there may be some extra parameters with image generator like number of images , negative prompts.

Comment (krokoko) [reply lambda/aws-qa-appsync-opensearch/question_answering/src/adapters/bedrock/titan.py]: addressed in [c4b5cb5](https://github.com/awslabs/generative-ai-cdk-constructs/pull/332/commits/c4b5cb5214fdc259dca30ed76b1d8d07348635b4)

File: lambda/aws-qa-appsync-opensearch/question_answering/src/qa_agent/chain.py [hunk 4]
```diff
@@ -95,21 +101,18 @@ def run_question_answering(arguments):
             return ''
 
         model_id=qa_model['modelId']
-        model_max_tokens = get_max_tokens(model_id)
+        model_provider=qa_model['provider']
+        model_max_tokens = get_max_tokens(model_provider+'.'+model_id)
```
Comment (dineshSajwan) [inline lambda/aws-qa-appsync-opensearch/question_answering/src/qa_agent/chain.py:105]: is this model_id coming from schema directly or is mapped with bedrockModel. from schema the user is expected to enter "anthropic.claude-3-sonnet-20240229-v1:0" and the get_max_token expect it to be BedrockModel.ANTHROPIC_CLAUDE_V3_SONNET_V1

**Review by dineshSajwan — APPROVED** [review]

**Review by laithalsaadoon — APPROVED** [review]
👍

**Pull request merged**